Migration 11.0

guewen opened the pull request:
Will move to https://github.com/OCA/ddmrp/pull/3

JordiBForgeFlow committed:
change supply qualification for net flow position to consider only supply
the moves that are within the DLT.

Commit:
Simplify orderpoint_id fields on purchase and mrp.production

This is now a standard Many2one field in the purchase addon.
Keep the index though.

For mrp.production, we add the field which is filled on creation of the
manufacturing order.

HviorForgeFlow committed:
adding ddmrp buffer profiles data

Commit:
Migrate to model / api changes done in Odoo 11.0

* Move method override to the new wizard
  procure_calculation has been moved to a wizard working on procurement.group
  instead of procurement.order
* Adapt name of report model renamed in core mrp
* Use correct format in views for colors
* Update ir_cron xml to new structure
* Rename field product_uom_id on mrp.bom

LoisRForgeFlow committed:
[FIX] setup and travis

LoisRForgeFlow committed:
[10.0][MIG] ddmrp_adjustment: migrate latest changes

Commit:
Remove procurement.order

Commit:
Fix typos in readme

Commit:
Fix tests

And use a SavepointCase which reduces the tests from ~10 seconds as the
setup is done only once

LoisRForgeFlow committed:
[10.0][FIX] ddmrp: enable the use of override buffers

guewen commented:
Travis is not running the tests because they don't pass at this point :laughing:

LoisRForgeFlow committed:
[9.0][FIX] ddmrp: issue with date on manually created pickings.

Commit:
[10.0][ADD] Execution history buffer (#49)

guewen commented:
Rebased from 10.0. All tests pass!

guewen commented on ddmrp/models/procurement_rule.py:
```diff
@@ -0,0 +1,33 @@
+        )
+        orderpoint = values.get('orderpoint_id')
+        if orderpoint:
+            orderpoint.cron_actions()
```
This was not done in 9.0 and 10.0. I'm not sure it's good performance-wise. It was necessary for the test `test_buffer_zones_all` to pass after the generation of the `mrp.production` otherwise the stats would not be up-to-date and the recommended quantity would still be 40 instead of 0 (in the last check)

simahawk replied:
if ít's just for tests wouldn't be better to state it w/ a comment and use context key to enable it on demand (and use it tests)

guewen replied:
I don't think it's just for tests, it was just that I spotted the requirement to call it with an existing test which was failing.

LoisRForgeFlow replied:
I think it's good to refresh the buffer/orderpoint after a MO is created, so the MO execution status will be updated too. 

Ideal would be to handle all the updates using the OCA/queue, not in the migration scope though.

guewen commented on ddmrp/models/stock_warehouse_orderpoint.py:
```diff
@@ -619,6 +645,7 @@ def cron_ddmrp(self, automatic=False):
         orderpoints = self.search([])
         i = 0
         j = len(orderpoints)
+        orderpoints.refresh()
```
Having the refresh here instead of in each `_calc` method reduces the `test_procure_recommended` from ~8s to ~4s on my machine. Tests still pass but I'm not 100% confident there is no side effect.

LoisRForgeFlow replied:
I think there is no side effect, good for me

guewen commented:
Proposed on https://github.com/OCA/ddmrp/pull/3 but I'm not sure if we want to work here first and extract only when we're sure we're good or move early to OCA. As there are changes on both 10.0 and 11.0 these days, I think it'd be easier to only work here and open the OCA PR later.

LoisRForgeFlow committed:
[10.0][FIX] ddmrp: dlt computeation considers distributed parts.

grindtildeath commented on ddmrp/data/ir_cron.xml:
```diff
@@ -10,9 +10,8 @@
         <field name="numbercall">-1</field>
         <field name="priority">6</field>
         <field eval="False" name="doall"/>
```
Please add `<field name="state">code</field>`

guewen replied:
Oops :sweat: I added them but removed them afterwards. Weird it's not the default value.
Fixed now

grindtildeath commented on ddmrp/data/ir_cron.xml:
```diff
@@ -26,9 +25,8 @@
         <field name="mutually_exclusive_cron_ids"
                eval="[(6, 0, [ref('ir_cron_ddmrp_calculation_scheduler_action'),])]"/>
         <field eval="False" name="doall"/>
```
Please add `<field name="state">code</field>`

grindtildeath reviewed (changes requested):
Still testing, but without this, the cron jobs do nothing actually.

Commit:
Compute adu value on create and write

The adu is computed by an onchange, but as the field is readonly, the
value is never passed to the server so doesn't change. Force the
computation on write and create.

Commit:
Use the buffer profile which match with the test data

guewen commented:
I rebased from 10.0 to get the last changes (https://github.com/Eficent/ddmrp/commit/a88a005d2e3aae1a4fd664243e58ce691dc94b21, ...)

grindtildeath commented on ddmrp/models/stock_warehouse_orderpoint.py:
```diff
@@ -315,6 +337,19 @@ def _compute_dlt(self):
+
     @api.multi
     @api.onchange("red_zone_qty")
     def onchange_red_zone_qty(self):
```
`red_zone_qty` is computed, but its computation doesn't trigger systematically the onchange.
In case the orderpoint `adu_calculation_method` is Fixed, and we change `fixed_adu`, the onchange is played correctly in the form view.
However, if we switch `adu_calculation_method` to Past, it's only when saving that `adu` is recomputed as is `red_zone_qty`. In this case the onchange isn't triggered and `product_min_qty` stays the same as it was when defined with fixed ADU.

The same behaviour applies to `onchange_green_zone_qty` at line 368.

LoisRForgeFlow replied:
It's recommended to keep ADU computed daily and all the other parameters of the buffer the closest to real time possible. Knowing that, I wouldn't trigger the update of any value if ADU is detected to change (fixed adu or any other thing that related to ADU) but wait until the next daily update.

grindtildeath commented on ddmrp/models/stock_warehouse_orderpoint.py:
```diff
@@ -371,20 +406,6 @@ def open_moves(self):
-        else:
-            return 0.0
-
     @api.model
```
This should be `@api.multi`
Same at lines 424, 453, 460, 372.
I didn't check other classes yet.

hparfr committed:
Use expected_date instead of date in all cacl_qualified_deman

grindtildeath committed:
Ensure _calc_adu is called to play onchange on product_min/max_qty

grindtildeath committed:
Fix api decorators and add ensure_one()

HviorForgeFlow committed:
Merge pull request #55 from akretion/fix/date_expcted_in_calc_qualified_demand

[10][FIX] Use expected_date instead of date in all cacl_qualified_deman

rousseldenis committed:
[10.0][FIX] ddmrp: Change self to rec to avoid Singleton errors

simahawk commented on ddmrp/models/procurement_group.py:
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
```
remove coding from *.py

simahawk commented on ddmrp/models/procurement_rule.py:
```diff
@@ -0,0 +1,33 @@
+            product_id, product_qty, product_uom,
+            location_id, name, origin, values
+        )
+        orderpoint = values.get('orderpoint_id')
```
aren't you changing `orderpoint_id` from recordset to ID right above?

grindtildeath replied:
The change above is unrelated to the values we get here :
```
> /opt/odoo/external-src/ddmrp/ddmrp/models/procurement_rule.py(30)_run_manufacture()
-> orderpoint = values.get('orderpoint_id')
(Pdb) values
{'group_id': procurement.group(), 'date_planned': '2018-04-10 11:38:37', 'orderpoint_id': stock.warehouse.orderpoint(119,), 'company_id': res.company(1,), 'warehouse_id': stock.warehouse(1,), 'priority': '1'}
```

simahawk commented on ddmrp/models/stock_warehouse_orderpoint.py:
```diff
@@ -315,6 +337,19 @@ def _compute_dlt(self):
+    @api.multi
+    def write(self, vals):
+        super(StockWarehouseOrderpoint, self).write(vals)
+        if not self.env.context.get('__calc_adu'):
```
can we make this ctx key more meaningful? I don't expect `__calc_adu` to disable the calc :)

simahawk commented on ddmrp/views/mrp_production_view.xml:
```diff
@@ -8,11 +8,13 @@
                 <field name="execution_priority_level"
-                       bg_color="red:execution_priority_level=='1_red';yellow:execution_priority_level=='2_yellow';green:execution_priority_level == '3_green'"
-                       fg_color="white:execution_priority_level=='1_red';black:execution_priority_level=='2_yellow';white:execution_priority_level == '3_green'"/>
+                       options="{
```
would be better to use single quote to wrap and double quotes inside, to make this JSON compat OOTB

guewen replied:
It's not JSON but a python dict evaluated with pyeval so not sure its useful?

simahawk commented on ddmrp_history/models/stock_warehouse_orderpoint.py:
```diff
@@ -112,8 +122,113 @@ def stacked(df, categories):
+            last = np.zeros(len(df[categories[0]]))
+            last += start_stack
+            for cat in categories:
+                next = last + df[cat]
```
use `_next` to not clash w/ built-ins

grindtildeath replied:
Attended here : https://github.com/Eficent/ddmrp/pull/52/commits/cb36fca16a87a1623e522c03917af78a587eb4c1

simahawk commented on ddmrp_history/models/stock_warehouse_orderpoint.py:
```diff
@@ -112,8 +122,113 @@ def stacked(df, categories):
+        start_stack = 0
+
+        def stacked(df, categories):
+            areas = dict()
```
`areas = {}`

grindtildeath replied:
Attended here : https://github.com/Eficent/ddmrp/pull/52/commits/cb36fca16a87a1623e522c03917af78a587eb4c1

simahawk commented on ddmrp_history/models/stock_warehouse_orderpoint.py:
```diff
@@ -112,8 +122,113 @@ def stacked(df, categories):
+    def _compute_execution_history_chart(self):
+        start_stack = 0
+
+        def stacked(df, categories):
```
docstring?

simahawk commented on ddmrp_history/models/stock_warehouse_orderpoint.py:
```diff
@@ -112,8 +122,113 @@ def stacked(df, categories):
+                last = next
+            return areas
+
+        for rec in self:
```
save some space by declaring at beginning of the loop:
```python
history_model = self.env['ddmrp.history']
domain = [('orderpoint_id', '=', rec.id)]
```

grindtildeath replied:
Attended here : https://github.com/Eficent/ddmrp/pull/52/commits/cb36fca16a87a1623e522c03917af78a587eb4c1

simahawk commented on ddmrp_history/models/stock_warehouse_orderpoint.py:
```diff
@@ -112,8 +122,113 @@ def stacked(df, categories):
-    # TODO: computed through a button? able to select period.
-    history_chart = fields.Text(
-        string='Historical Chart', compute='_compute_history_chart')
+    def _compute_execution_history_chart(self):
```
docstring

simahawk commented on ddmrp_history/models/stock_warehouse_orderpoint.py:
```diff
@@ -112,8 +122,113 @@ def stacked(df, categories):
+                          'top_of_yellow', 'top_of_red', 'dark_red']
+            data = {}
+
+            dates = [datetime.strptime(
```
some comments below here might help to understand what's going on

LoisRForgeFlow commented:
Just code reviewed the tests and they look good. Just one comment: `dlt` should be removed from all the creating dicts for buffers as it is now computed, e.g. [example](https://github.com/guewen/ddmrp/blob/ee1710ee3d14bf2b57d2216673c493dbc0905915/ddmrp/tests/test_ddmrp.py#L194)

Commit:
Merge remote-tracking branch 'eficent/10.0' into mig-ddmrp

LoisRForgeFlow commented:
@guewen There are conflicts here, could you have a look?

grindtildeath committed:
Fix api decorator on create of mrp.production

grindtildeath committed:
Remove coding: utf-8 declaration from *.py files

grindtildeath committed:
Rename context key __calc_adu to __no_adu_calc

grindtildeath committed:
Remove dlt from buffer creation dicts in tests

grindtildeath commented:
@guewen @simahawk
I attended most of the changes asked.
@lreficent 
Don't know if the conflicts you mentioned have been solved in the meanwhile, but it looks ok to me at the moment.

LoisRForgeFlow commented:
@grindtildeath ~the conflicts haven't been solved, I cannot merge due to them~

EDIT: Nevermind, I was to quick answering. I can use any of the other methods to merge :smile:

LoisRForgeFlow commented:
I will try to review and merge this week.

LoisRForgeFlow committed:
[11.0][MIG] ddmrp:
* bump version
* remove unneded dependency
* fix small issues on views
* fix procure_recommended_qty computation

LoisRForgeFlow commented:
@guewen I proposed you a PR: https://github.com/guewen/ddmrp/pull/1

Once you merge that one I think we are ready to move forward.

Commit:
Merge pull request #1 from Eficent/11.0-guewen

[11.0] ddmrp migration final fixes

guewen commented:
@lreficent merged, thanks

Commit:
Fix lint

LoisRForgeFlow commented:
Finally merged! Thank you all for the effort! :smile:

guewen commented:
@lreficent should I update https://github.com/OCA/ddmrp/pull/3 or is it better to wait we have all the addons?

LoisRForgeFlow commented:
@guewen I think there is only one addon pending to merge in OCA ([mrp_bom_location](https://github.com/OCA/manufacture/pull/254), it will be only a rebase when it is merged, but up to you!

guewen commented:
I was speaking about #52, #51 and #50, maybe it's simpler to move to OCA once they have been merged as well

LoisRForgeFlow commented:
Oh, yes, I think if you rebase them I can review and merge here next week.

JordiBForgeFlow commented:
@guewen I agree that perhaps in this case it is better to send a single PR for all the modules in DDMRP. We know all the reviewers invoved (maxime, joël, us, raphaël, etc...) and all are very familiar with the modules.
